### PR TITLE
[Form] [DoctrineBridge] optimized LazyChoiceList and DoctrineChoiceLoader

### PR DESCRIPTION
**Pull request by HeahDude**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | ~ |
# Problem

Actually we got a circular dependency:

| Object | Return |
| --- | --- |
| `DefaultChoiceListFactory::createListFromLoader()` (decorated) | `LazyChoiceList` with loader and resolved `$value` |
| `LazyChoiceList::get*` | `DoctrineChoiceLoader` with resolved `$value` |
| `DoctrineChoiceLoader::loadChoiceList()` | (decorated) `DefaultChoiceListFactory` with loaded choices and resolved `$value` |
| `DefaultChoiceListFactory::createListFromChoices()` | `ArrayChoiceList` with `resolved`$value` |

With this refactoring, the `DoctrineChoiceLoader` is no longer dependant to the factory, the `ChoiceLoaderInterface::loadChoiceList()` must return a `ChoiceListInterface` but should not need a decorated factory while `$value` is already resolved. It should remain lazy IMHO.
# Solution

| Object | Return |
| --- | --- |
| `DefaultChoiceListFactory::createListFromLoader()` | `LazyChoiceList` with loader and resolved `$value` |
| `LazyChoiceList::get*()` | `DoctrineChoiceLoader` with resolved `$value` |
| `DoctrineChoiceLoader::loadChoiceList()` | `ArrayChoiceList` with resolved `$value`. |

Since `choiceListFactory` is a private property, this change should be safe regarding BC.

To justify this change, I've made some blackfire profiling.
You can see my [branch of SE here](https://github.com/HeahDude/symfony-standard/tree/test/optimize-doctrine_choice_loader) and the [all in file test implementation](https://github.com/HeahDude/symfony-standard/blob/test/optimize-doctrine_choice_loader/src/AppBundle/Controller/DefaultController.php).

Basically it loads a form with 3 `EntityType` fields with different classes holding 50 instances each.

(INIT events are profiled with an empty cache)

| When | What | Diff (SE => PR) |
| --- | --- | --- |
| INIT (1) | build form (load types) | [see](https://blackfire.io/profiles/compare/061d5d28-15c6-4e01-b8c0-3edc9cb8daf0/graph) |
| INIT (2) | build view (load choices) | [see](https://blackfire.io/profiles/compare/04f142a8-d886-405a-be4d-636ba82d8acd/graph) |
| CACHED | build form (load types) | [see](https://blackfire.io/profiles/compare/293b27b6-aa58-42ae-bafb-655513201505/graph) |
| CACHED | build view (load choices) | [see](https://blackfire.io/profiles/compare/e5b37dfe-cc9e-498f-b98a-7448830ad190/graph) |
| SUBMIT | build form (load types) | [see](https://blackfire.io/profiles/compare/7f3baea9-0d27-46b6-8c24-c577742382dc/graph) |
| SUBMIT | handle request (load choices) | [see](https://blackfire.io/profiles/compare/8644ebfb-4397-495b-8f3d-1a6e1d7f8476/graph) |
| SUBMIT | build view (load values) | [see](https://blackfire.io/profiles/compare/89c3cc7c-ea07-4300-91b3-99004cb58ea1/graph) |

(1):
build_form-no_cache
(2):
build_view-no_cache

It can seem like 1 and 2 balance each other but it comes clear when comparing values:

| - | Build form | Build view |
| --- | --- | --- |
| wall time | -88 ms | +9.71 ms |
| blocking I/O | -40 ms | +3.67 ms |
| cpu | -48 ms | +13.4 ms |
| memory | -4.03 MB | +236 kB |
| network | -203 B | +2.21 kB |

### CODE REVIEW

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php [hunk 1]
```diff
@@ -69,15 +70,19 @@ class DoctrineChoiceLoader implements ChoiceLoaderInterface
...
-        $this->factory = $factory;
         $this->manager = $manager;
         $this->class = $classMetadata->getName();
         $this->idReader = $idReader ?: new IdReader($manager, $classMetadata);
         $this->objectLoader = $objectLoader;
+
+        if (null !== $this->factory) {
+            @trigger_error(sprintf('Passing a ChoiceListFactoryInterface to %s is deprecated since version 3.1 and will no longer be supported in 4.0. You should either call "%s::loadChoiceList" or override it to return a ChoiceListInterface.', __CLASS__, __CLASS__));
+            $this->factory = $factory;
+        }
```
Comment (iltar) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: This deprecation means that in the future versions it will always have to be called with null as first argument. It's a pain to move the argument to the end of the list but I think that's the only solution you can do for 4.0

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php [hunk 2]
```diff
@@ -146,7 +149,7 @@ public function loadChoicesForValues(array $values, $value = null)
 
         // Optimize performance in case we have an object loader and
         // a single-field identifier
-        $optimize = null === $value || is_array($value) && $value[0] === $this->idReader;
+        $optimize = is_array($value) && $this->idReader && $value[0];
```
Comment (stof) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: This change is broken

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php [hunk 3]
```diff
@@ -93,9 +98,7 @@ public function loadChoiceList($value = null)
             ? $this->objectLoader->getEntities()
             : $this->manager->getRepository($this->class)->findAll();
 
-        $this->choiceList = $this->factory->createListFromChoices($objects, $value);
-
-        return $this->choiceList;
+        return $this->choiceList = new ArrayChoiceList($objects, $value);
```
Comment (stof) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: This is broken, as it means that ``$value`` cannot be a property path anymore (as you don't handle the conversion to a closure anymore)

Comment (stof) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: btw, this also prevents re-using existing choice lists in the caching factory

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php (hunk 1 above)
Comment (stof) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: The proper deprecation way would be to allow moving all args to the left, not to allow passing null as same argument (i.e. allowing using the right Symfony 4 API)

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php (hunk 3 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: When the `ChoiceListFactory` create first a `LazyChoiceList` instance (the first and mandatory call) all the decorators get pass the `$value` so they can resolve it.

Once this step done the `LazyChoiceList` will be waken up to use its loader and resolved value all together.

The thing is actually instead of return an `ArrayChoiceList` instance with the loaded choices, the `DoctrineChoiceLoader` (which is the only `ChoiceLoaderInterface` implemented in core) get passed the hole factory with all the decorators (and might add some in the future as it is one a the few BC ways to add feature for now).

I think this final call should be delegated, that remains coherent with the interfaces and has a performance impact.

Anyway it does not break the decorator resolution.

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php (hunk 2 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: Why ? it's internal, it's always an array. But I realise that even before it does not test `asset($value[0])` but that's for 2.7

Comment (HeahDude) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: But I guess that if it's an array it is a callable array in any case, right ?

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php (hunk 1 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: Ok will do that, thanks guys for the review :)

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php (hunk 2 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: So we're missing a test :)

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php (hunk 3 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: The test with the property path added in #17694 is not broken (ref https://github.com/symfony/symfony/pull/17694/files#diff-641188f07a887e51cad48e9d72d69d07R754)

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php (hunk 2 above)
Comment (stof) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: the previous check was right. Here, you check that ``$value[0]`` is any truthy value, which does not make any sense (it will always be the case for a callable array anyway)

Comment (stof) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: and the condition ``$this->idReader`` will also always be truthy here

Comment (HeahDude) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: Ok my bad, I guess I read too much, until now I read at `is_array($value) && $this->idReader === $value[0];`... 

I'm gonna take a nap :palm_tree:

**Comment by HeahDude** [comment]
Comments addressed.

**Comment by HeahDude** [comment]
Actually the more relevant of all the profile is cached app and form types loading [profile 3](https://blackfire.io/profiles/compare/293b27b6-aa58-42ae-bafb-655513201505/graph?settings%5Bdimension%5D=wt&settings%5Bdisplay%5D=landscape&settings%5BtabPane%5D=nodes&selected=&callname=main()):
load_types-cached

**Comment by HeahDude** [comment]
It should also be the most used (just on display) with a warmup cache

**Comment by HeahDude** [comment]
This might be a first step to do better but can be improved even more IMHO, so let's do this for 3.2.

Status: Needs Work

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php [hunk 4]
```diff
@@ -60,20 +61,30 @@ class DoctrineChoiceLoader implements ChoiceLoaderInterface
...
      * @param string                     $class        The class name of the
      *                                                 loaded objects
      * @param IdReader                   $idReader     The reader for the object
      *                                                 IDs.
+     * @param ChoiceListFactoryInterface $factory      The factory for creating
+     *                                                 the loaded choice list
      * @param null|EntityLoaderInterface $objectLoader The objects loader
      */
-    public function __construct(ChoiceListFactoryInterface $factory, ObjectManager $manager, $class, IdReader $idReader = null, EntityLoaderInterface $objectLoader = null)
+    public function __construct($manager, $class, $idReader = null, $objectLoader = null, $factory = null)
```
Comment (webmozart) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: This is a BC break. We cannot reorder the arguments of the constructor.

Comment (webmozart) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: Ah I just realized what you are doing below.

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php [hunk 5]
```diff
@@ -60,20 +61,30 @@ class DoctrineChoiceLoader implements ChoiceLoaderInterface
...
     {
+        // BC to be removed in 4.0
+        if (null !== $this->factory && $objectLoader instanceof EntityLoaderInterface) {
```
Comment (webmozart) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: For clarity, I'd change the `instanceof` check to `$manager instanceof ChoiceListFactoryInterface`. I'd also add a comment that we need to add back the type hints in 4.0.

File: src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php [hunk 6]
```diff
@@ -160,11 +160,11 @@ public function configureOptions(OptionsResolver $resolver)
                 }
 
                 $doctrineChoiceLoader = new DoctrineChoiceLoader(
-                    $this->choiceListFactory,
                     $options['em'],
                     $options['class'],
                     $options['id_reader'],
                     $entityLoader
+                    // Passing a {@link ChoiceListFactory} instance is deprecated as of 3.1
```
Comment (webmozart) [inline src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php]: I think we don't need this comment here.

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php (hunk 4 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: Yes I'm just feel bad about removing the type hints but @iltar and @stof were suggesting that change for good reasons I guess.

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php (hunk 5 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: Sounds good!

**Comment by HeahDude** [comment]
Should it be done for 3.1 ?

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php [hunk 7]
```diff
@@ -146,7 +155,7 @@ public function loadChoicesForValues(array $values, $value = null)
 
         // Optimize performance in case we have an object loader and
         // a single-field identifier
-        $optimize = null === $value || is_array($value) && $value[0] === $this->idReader;
+        $optimize = is_array($value) && $this->idReader === $value[0];
```
Comment (webmozart) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: Why did you remove the `null === $value` check?

Comment (HeahDude) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: Because I read it as `null === $value && is_array($value)` :disappointed:

**Comment by HeahDude** [comment]
Comments addressed.

Status: Needs Review

**Comment by HeahDude** [comment]
Should I deprecate the caching of `LazyChoiceList` in that PR too ? (ref https://github.com/symfony/symfony/pull/18332#issuecomment-203779109)

**Comment by HeahDude** [comment]
@javiereguiluz I guess this one could have a "Enhancement" label :)

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php (hunk 7 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: Even if I was the one "fixing" it in the first place and introducing a regression at the same type...

However it's clear to me now that `choice_value` in entity type should better not be used on big collections to keep this optimisation efficient.

What I think of now is that actually `$value` is always resolved by factories when stored in the cached `LazyChoiceList` which passes it to the loader each time.

What about creating another `ChoiceListInterface` depending on this logic while moving it in a `Symfony\Bridge\Doctrine\Form\ChoiceList\Factory\SingleIdFactoryDecorator` ?

Do you think it's doable to cache a map `id => value` when `$value` is not null somewhere ?

What about possibilities added by the new Cache component ?
ids should be immutable, right ? But their corresponding values may need to be updated but must not change during the lifecycle of a form since we rely on it to select choices.

How to generate a good hash based on the class and `$value` in that case ? that should not be a problem if `$value` is a string property path but if it is a closure ?

One way I see this could be implemented is by adding a `Symfony\Bridge\Doctrine\Form\ChoiceList\LazyEntityList` which would be basically a `LazyChoiceList` holding the `idReader` for mapping and using `$value` as well.

1. Current WorkFlow
-------------
1. Decorated factory creates and stores the `LazyChoiceList` storing the `DoctrineChoiceLoader` and  `$value = null`
2. `LazyChoiceList` calls the loader and passes `null` as value.
3. `DoctrineChoiceLoader` resolves `$value` with its `idReader` and load the choices
4. `DoctrineChoiceLoader` calls the decorated factory with resolved value and loaded choices
5. Decorated factory creates and stores the `ArrayChoiceList` will never call it back.
6. `DoctrineChoiceLoader` stores the `ArrayChoiceList` and will never call it back.
7. `LazyChoiceList` stores the `ArrayChoiceList` while storing the loader.

Also, if another `EntityType` is used in the same form, another `DoctrineChoiceLoader` is created by the `DoctrineType` and the same process happens again.

When the same `em`, `class` and 'query_builder` options are used for the `EntityType` the same loader is return, so the factory returns the same `LazyChoiceList` storing the loaded `ArrayChoiceList`.

2. Workflow a bit improved by this PR

Removes line 5 from above.

3. Ideal Workflow
---------
1. IdReader decorated factory resolves `$value`, `$idReader` and `$entityLoader` (instead of DoctrineType) then creates a `LazyEntityList` storing the `DoctrineChoiceLoader` and the resolved arguments, 
2. `LazyEntityList` calls the loader and passes it the resolved arguments.
3. `DoctrineChoiceLoader` load the choices with optimisation
4. `DoctrineChoiceLoader` creates an `EntityList` with an extra map ids by values
5. `DoctrineChoiceLoader` stores the `EntityList`.

What do you think ? Thanks.

**Comment by HeahDude** [comment]
I pushed the `LazyChoiceList` optimization, with both it's even better, here some profiles from previous optimization of the loader to both loader and LazyChoiceList with this new [commit](:

 - [no cache](https://blackfire.io/profiles/compare/2b99d194-8808-4b85-81cb-2e677b807709/graph)
 - [with cache](https://blackfire.io/profiles/compare/5366b0a0-dbb0-412e-a2e5-267fe71457e8/graph)
 - [on submit](https://blackfire.io/profiles/compare/faadae3e-f987-442d-8192-5447f882ecad/graph)

So with both optimizations from current to 4.0 (without BC layer):

 - [no cache](https://blackfire.io/profiles/compare/9dfdf55a-8b3a-40f8-982e-b5436d630064/graph)

    full_optimized-no_cache

 - [with cache](https://blackfire.io/profiles/compare/abeb8d8c-64de-4efe-9ef2-4a15377b094a/graph)

    full_optimized-cached

 - [on submit 3 fields with one choice of 50 for each](https://blackfire.io/profiles/compare/8548d855-5631-493d-acfe-342dfad7d7f6/graph)

    full_optimized-on_submit

Finally profiles with deprecation layer (the final look after that PR):

- 3 entity fields with 50 entities each:

  - [no cache](https://blackfire.io/profiles/compare/85334c69-1dc2-4dfd-8bef-af1327149cd7/graph)

  - [cache](https://blackfire.io/profiles/compare/b7f1e75e-40ce-40cc-9d8e-979b77236c5c/graph)

  - [submit](https://blackfire.io/profiles/compare/871ebdd4-4e56-465d-9283-9ed5dcca744e/graph)

- 1 entity field with 150 entities:

  - [display](https://blackfire.io/profiles/compare/e089a246-7cd7-4ce7-bc3c-69cc72c8fbea/graph)

  - [submit](https://blackfire.io/profiles/compare/34013bc6-a3ee-477b-bfcf-031d068e61e3/graph)

And that will be even better in 4.0!

Don't forget to look at the real values, they are more impacted than the percentages :)

Note that actually the code as it works by default has not triggered any deprecations for my tests, it should only happen for custom implementations and that's good.

**Comment by HeahDude** [comment]
Failures are unrelated.

**Comment by javiereguiluz** [comment]
@HeahDude nice comparisons :clap: And thanks for using realistic benchmarks (it's "easy" to improve performance when doing a benchmark with 1 million entities ... but that's no real)

**Comment by HeahDude** [comment]
Thanks @javiereguiluz, still I guess it would be nice if someone could test it in a real app too to confirm it does not mess with anything...

**Comment by HeahDude** [comment]
What's the status of this PR? is this still to be in 3.1?

Allied to #18332 we can get a great performance improvement.

I tested it with `symfony-demo` by quickly adding some `EntityType` fields in some forms, and it worked like a charm :)

Anyone for a final review?

**Comment by fabpot** [comment]
Thank you @HeahDude.
